File: src/AutoContext.Framework/Transport/PipeStreamingClient.cs
namespace AutoContext.Framework.Transport;

using System.IO.Pipes;
using System.Threading.Channels;

using Microsoft.Extensions.Logging;

/// <summary>
/// Layer-3 streaming client: drains items from a bounded queue and
/// writes their serialized bytes over a named pipe. Drop-oldest queue
/// semantics keep callers non-blocking. On any I/O failure the stream
/// is closed and remaining items (and any future <see cref="Post"/>
/// items) are routed to <paramref name="fallback"/>.
/// </summary>
/// <remarks>
/// No reconnect policy is built in — matches today's "logger of last
/// resort" behavior in <c>LoggingClient</c>. A dedicated drain task
/// owns all I/O; the type itself is thread-safe for <see cref="Post"/>.
/// Designed to be wrapped by an endpoint class (e.g.
/// <c>LoggingClient</c>) that supplies the <typeparamref name="T"/>
/// type, the serializer, and the fallback path.
/// </remarks>
public sealed partial class PipeStreamingClient<T> : IAsyncDisposable
{
    private const int DefaultDrainTimeoutMs = 2000;

    private readonly PipeTransport _transport;
    private readonly string _pipeName;
    private readonly PipeDirection _direction;
    private readonly int _connectTimeoutMs;
    private readonly ReadOnlyMemory<byte> _greeting;
    private readonly Func<T, ReadOnlyMemory<byte>> _serialize;
    private readonly Action<T>? _fallback;
    private readonly ILogger<PipeStreamingClient<T>> _logger;
    private readonly Channel<T> _queue;
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _drainTask;
    private int _disposed;

    /// <summary>
    /// Creates and starts a new streaming client. The drain task runs
    /// until <see cref="DisposeAsync"/> is called.
    /// </summary>
    /// <param name="transport">Connect primitive used by the drain task.</param>
    /// <param name="pipeName">Pipe name; pass empty to disable the pipe and
    /// route every <see cref="Post"/> item through <paramref name="fallback"/>.</param>
    /// <param name="serialize">Maps an item to the bytes written to the pipe.
    /// Called on the drain task; should not allocate excessively.</param>
    /// <param name="logger">Required logger for diagnostic output.</param>
    /// <param name="greeting">Optional handshake bytes written immediately
    /// after connect, before any item is drained. Empty means "skip greeting".</param>
    /// <param name="fallback">Optional sink invoked when the pipe is
    /// unavailable or has broken. Called on the drain task.</param>
    /// <param name="direction">Pipe direction; defaults to <see cref="PipeDirection.Out"/>.</param>
    /// <param name="queueCapacity">Maximum buffered items. Drop-oldest when full.</param>
    /// <param name="connectTimeoutMs">Connect timeout in milliseconds.</param>
    public PipeStreamingClient(
        PipeTransport transport,
        string pipeName,
        Func<T, ReadOnlyMemory<byte>> serialize,
        ILogger<PipeStreamingClient<T>> logger,
        ReadOnlyMemory<byte> greeting = default,
        Action<T>? fallback = null,
        PipeDirection direction = PipeDirection.Out,
        int queueCapacity = 1024,
        int connectTimeoutMs = 2000)
    {
        ArgumentNullException.ThrowIfNull(transport);
        ArgumentNullException.ThrowIfNull(pipeName);
        ArgumentNullException.ThrowIfNull(serialize);
        ArgumentNullException.ThrowIfNull(logger);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(queueCapacity);

        _transport = transport;
        _pipeName = pipeName;
        _direction = direction;
        _connectTimeoutMs = connectTimeoutMs;
        _greeting = greeting;
        _serialize = serialize;
        _fallback = fallback;
        _logger = logger;
        _queue = Channel.CreateBounded<T>(
            new BoundedChannelOptions(queueCapacity)
            {
                FullMode = BoundedChannelFullMode.DropOldest,
                SingleReader = true,
                SingleWriter = false,
            });
        _drainTask = Task.Run(() => DrainAsync(_cts.Token));
    }

    /// <summary>
    /// Posts <paramref name="item"/> for off-thread delivery. Never
    /// blocks; if the queue is full the oldest entry is dropped. Returns
    /// <see langword="false"/> if the client has been disposed.
    /// </summary>
    public bool Post(T item) => _queue.Writer.TryWrite(item);

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
        {
            return;
        }

        _queue.Writer.TryComplete();

        try
        {
            await _drainTask.WaitAsync(TimeSpan.FromMilliseconds(DefaultDrainTimeoutMs)).ConfigureAwait(false);
        }
        catch (TimeoutException)
        {
            // Drain didn't finish in time — abandon it.
        }
        catch (OperationCanceledException)
        {
            // Expected during shutdown.
        }

        await _cts.CancelAsync().ConfigureAwait(false);

        if (_drainTask.IsCompleted)
        {
            _cts.Dispose();
        }
    }

    private async Task DrainAsync(CancellationToken cancellationToken)
    {
        Stream? stream = await TryConnectAsync(cancellationToken).ConfigureAwait(false);

        try
        {
            if (stream is not null && _greeting.Length > 0
                && !await TryWriteAsync(stream, _greeting, cancellationToken).ConfigureAwait(false))
            {
                await stream.DisposeAsync().ConfigureAwait(false);
                stream = null;
            }

            await foreach (var item in _queue.Reader.ReadAllAsync(cancellationToken).ConfigureAwait(false))
            {
                if (stream is not null)
                {
                    var bytes = _serialize(item);
                    if (await TryWriteAsync(stream, bytes, cancellationToken).ConfigureAwait(false))
                    {
                        continue;
                    }

                    LogStreamBroken(_logger, _pipeName);
                    await stream.DisposeAsync().ConfigureAwait(false);
                    stream = null;
                }

                _fallback?.Invoke(item);
            }
        }
        catch (OperationCanceledException)
        {
            // Shutdown — fall through.
        }
        finally
        {
            if (stream is not null)
            {
                await stream.DisposeAsync().ConfigureAwait(false);
            }
        }
    }

    private async Task<Stream?> TryConnectAsync(CancellationToken cancellationToken)
    {
        if (string.IsNullOrEmpty(_pipeName))
        {
            return null;
        }

        try
        {
            return await _transport.ConnectAsync(_pipeName, _connectTimeoutMs, _direction, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is TimeoutException or IOException or UnauthorizedAccessException or OperationCanceledException)
        {
            return null;
        }
    }

    private static async Task<bool> TryWriteAsync(Stream stream, ReadOnlyMemory<byte> bytes, CancellationToken cancellationToken)
    {
        try
        {
            await stream.WriteAsync(bytes, cancellationToken).ConfigureAwait(false);
            await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
            return true;
        }
        catch (Exception ex) when (ex is IOException or ObjectDisposedException)
        {
            return false;
        }
    }

    [LoggerMessage(EventId = 1, Level = LogLevel.Debug,
        Message = "Streaming pipe '{PipeName}' broke; routing remaining items to fallback.")]
    private static partial void LogStreamBroken(ILogger logger, string pipeName);
}

File: src/AutoContext.Nodejs.Core/src/pipes/pipe-streaming-client.ts
import type { Socket } from 'node:net';

import type { LoggerFacade } from '../logging/logger-facade.js';
import { PipeTransport } from './pipe-transport.js';

/**
 * Layer-3 streaming client: drains items from a bounded queue and
 * writes their serialized bytes over a named pipe. Drop-oldest queue
 * semantics keep callers non-blocking. On any I/O failure the stream
 * is closed and remaining items (and any future {@link post} items)
 * are routed to {@link PipeStreamingClientOptions.fallback}.
 *
 * No reconnect policy is built in — matches today's "logger of last
 * resort" behavior in `LoggingClient`. Designed to be wrapped by an
 * endpoint class that supplies the {@link T} type, the serializer,
 * and the fallback path.
 *
 * Counterpart of the C# `PipeStreamingClient<T>` in
 * `AutoContext.Framework`.
 */
export interface PipeStreamingClientOptions<T> {
    readonly transport: PipeTransport;
    readonly pipeName: string;
    readonly serialize: (item: T) => Buffer | Uint8Array;
    readonly logger: LoggerFacade;
    readonly greeting?: Buffer | Uint8Array;
    readonly fallback?: (item: T) => void;
    readonly queueCapacity?: number;
    readonly connectTimeoutMs?: number;
}

export class PipeStreamingClient<T> {
    private static readonly DEFAULT_QUEUE_CAPACITY = 1024;
    private static readonly DEFAULT_CONNECT_TIMEOUT_MS = 2000;
    private static readonly DEFAULT_DRAIN_TIMEOUT_MS = 2000;

    private readonly transport: PipeTransport;
    private readonly pipeName: string;
    private readonly serialize: (item: T) => Buffer | Uint8Array;
    private readonly fallback: ((item: T) => void) | undefined;
    private readonly greeting: Buffer | Uint8Array | undefined;
    private readonly logger: LoggerFacade;
    private readonly queueCapacity: number;
    private readonly connectTimeoutMs: number;
    private readonly queue: T[] = [];
    private readonly waiters: Array<() => void> = [];
    private completed = false;
    private socket: Socket | undefined;
    private readonly drainTask: Promise<void>;
    private disposeTask: Promise<void> | undefined;

    constructor(opts: PipeStreamingClientOptions<T>) {
        this.transport = opts.transport;
        this.pipeName = opts.pipeName;
        this.serialize = opts.serialize;
        this.fallback = opts.fallback;
        this.greeting = opts.greeting;
        this.logger = opts.logger;
        this.queueCapacity = opts.queueCapacity ?? PipeStreamingClient.DEFAULT_QUEUE_CAPACITY;
        this.connectTimeoutMs = opts.connectTimeoutMs ?? PipeStreamingClient.DEFAULT_CONNECT_TIMEOUT_MS;
        this.drainTask = this.drain();
    }

    /**
     * Enqueues `item` for off-thread delivery. Never blocks; if the
     * queue is full the oldest entry is dropped.
     */
    post(item: T): void {
        if (this.completed) {
            return;
        }
        if (this.queue.length >= this.queueCapacity) {
            this.queue.shift();
        }
        this.queue.push(item);
        const waiter = this.waiters.shift();
        if (waiter !== undefined) {
            waiter();
        }
    }

    /**
     * Stops accepting items, waits up to two seconds for the drain
     * loop to flush, then tears the socket down. Safe to call multiple
     * times — concurrent callers share a single completion promise.
     */
    async dispose(): Promise<void> {
        this.disposeTask ??= this.disposeCore();
        return this.disposeTask;
    }

    private async disposeCore(): Promise<void> {
        this.completed = true;
        const waiters = this.waiters.splice(0);
        for (const waiter of waiters) {
            waiter();
        }

        let timer: NodeJS.Timeout | undefined;
        const timeout = new Promise<void>((resolve) => {
            timer = setTimeout(resolve, PipeStreamingClient.DEFAULT_DRAIN_TIMEOUT_MS);
            timer.unref();
        });
        try {
            await Promise.race([this.drainTask, timeout]);
        }
        finally {
            if (timer !== undefined) {
                clearTimeout(timer);
            }
        }

        const socket = this.socket;
        this.socket = undefined;
        if (socket !== undefined) {
            socket.destroy();
        }
    }

    private async drain(): Promise<void> {
        let socket = await this.tryConnect();

        if (socket !== null) {
            // Swallow late socket errors — without this the process crashes
            // on EPIPE / peer close because `PipeTransport.connect` removes
            // its own error listener once the connection is established.
            socket.on('error', () => { /* see comment */ });

            if (this.greeting !== undefined && this.greeting.length > 0
                && !await PipeStreamingClient.tryWrite(socket, this.greeting)) {
                socket.destroy();
                socket = null;
            }
        }
        this.socket = socket ?? undefined;

        try {
            while (true) {
                const item = await this.dequeue();
                if (item === undefined) {
                    return;
                }
                if (this.socket !== undefined) {
                    const bytes = this.serialize(item);
                    if (await PipeStreamingClient.tryWrite(this.socket, bytes)) {
                        continue;
                    }
                    this.logger.debug(`Pipe stream to '${this.pipeName}' broken; routing further items to fallback.`);
                    this.socket.destroy();
                    this.socket = undefined;
                }
                if (this.fallback !== undefined) {
                    this.fallback(item);
                }
            }
        }
        finally {
            const s = this.socket;
            this.socket = undefined;
            if (s !== undefined) {
                s.destroy();
            }
        }
    }

    private dequeue(): Promise<T | undefined> {
        if (this.queue.length > 0) {
            return Promise.resolve(this.queue.shift());
        }
        if (this.completed) {
            return Promise.resolve(undefined);
        }
        return new Promise<T | undefined>((resolve) => {
            this.waiters.push(() => {
                if (this.queue.length > 0) {
                    resolve(this.queue.shift());
                    return;
                }
                resolve(undefined);
            });
        });
    }

    private async tryConnect(): Promise<Socket | null> {
        if (this.pipeName === '') {
            return null;
        }
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this.connectTimeoutMs);
        timer.unref();
        try {
            return await this.transport.connect(this.pipeName, controller.signal);
        }
        catch {
            return null;
        }
        finally {
            clearTimeout(timer);
        }
    }

    private static tryWrite(socket: Socket, bytes: Buffer | Uint8Array): Promise<boolean> {
        return new Promise<boolean>((resolve) => {
            if (socket.destroyed || !socket.writable) {
                resolve(false);
                return;
            }
            socket.write(bytes, (err) => {
                resolve(err === null || err === undefined);
            });
        });
    }
}
